crates/sui-transaction-builder/src/lib.rs

```diff
@@ -97,6 +97,79 @@ impl TransactionBuilder {
         }
     }
 
+    /// Construct the transaction data for a dry run (uses min gas budget)
+    pub async fn tx_data_for_dry_run(
+        &self,
+        sender: SuiAddress,
+        kind: TransactionKind,
+        gas_budget: u64,
+        gas_price: u64,
+        gas_payment: Option<Vec<ObjectID>>,
+        gas_sponsor: Option<SuiAddress>,
+    ) -> TransactionData {
+        let gas_payment = self
+            .input_refs(gas_payment.unwrap_or_default().as_ref())
+            .await
+            .unwrap_or_else(|_| vec![]);
+        let gas_sponsor = gas_sponsor.unwrap_or(sender);
+        TransactionData::new_with_gas_coins_allow_sponsor(
+            kind,
+            sender,
+            gas_payment,
+            gas_budget,
+            gas_price,
+            gas_sponsor,
+        )
+    }
+
+    /// Construct the transaction data from a transaction kind and a few needed
+    /// Use the `input_coins` for pay_sui
+    pub async fn tx_data(
+        &self,
+        sender: SuiAddress,
+        kind: TransactionKind,
+        // this can be empty, use it for pay or pass vec![] for pay_sui
+        input_coins: Vec<ObjectID>,
+        gas_budget: u64,
+        gas_price: u64,
+        gas_payment: Option<Vec<ObjectID>>,
+        gas_sponsor: Option<SuiAddress>,
+    ) -> Result<TransactionData, anyhow::Error> {
+        let input_gas = if let Some(ref gas_payment) = gas_payment {
+            gas_payment.first()
+        } else {
+            None
+        };
+        let gas_payment = self
+            .select_gas(
+                sender,
+                input_gas.copied(),
+                gas_budget,
+                input_coins,
+                gas_price,
+            )
+            .await?;
```

Comment:
I don't understand what this is doing, it seems like if you explicitly asked to pay for a transaction using multiple coins, it could potentially ignore the coins you selected after the first one and pick new ones from your wallet? That seems like the wrong thing to do.

Reply:
I am not sure what's unclear here... The overall logic is a bit convoluted, I agree.

The `select_gas` function will use the `input_gas` if it is passed. If it is not passed, then it will get all the owned objects and try to find a gas coin that is not part of the `input_coins`.

Reply:
Looking at how this actually gets used, it makes me think that the CLI doesn't actually support gas smashing, and this function is only ever given zero or one gas coin. If that's the case, then the right thing to do is to change the signature so that it doesn't seem like it can take multiple coins only to ignore all but the first one, but if that's not the case, it means the CLI claims to support gas smashing, but doesn't, which is a bug.

Reply:
See https://github.com/MystenLabs/sui/pull/16947#issuecomment-2065049716